include/ck_tile/host/host_tensor.hpp

```diff
@@ -271,253 +291,597 @@ CK_TILE_HOST auto make_ParallelTensorFunctor(F f, Xs... xs)
+        assert(get_element_space_size() <= data.size());
+        mData = data;
     }
 
-    T& operator()(std::vector<std::size_t> idx)
+    auto& get_slicer(size_type dim) { return mSlicers[dim]; }
+
+    auto& get_slicer(size_type dim) const { return mSlicers[dim]; }
+
+    auto get_slicers()
     {
-        return mData[mDesc.GetOffsetFromMultiIndex(idx)];
+        using std::begin, std::next;
+
+        return iterator_range(begin(mSlicers), next(begin(mSlicers), get_num_of_dimension()));
     }
 
-    const T& operator()(std::vector<std::size_t> idx) const
+    auto get_slicers() const
     {
-        return mData[mDesc.GetOffsetFromMultiIndex(idx)];
+        using std::begin, std::next;
+
+        return iterator_range(begin(mSlicers), next(begin(mSlicers), get_num_of_dimension()));
     }
 
-    typename Data::iterator begin() { return mData.begin(); }
+    private:
+    template <typename Self>
+    static decltype(auto) get_impl(Self&& self, span<const size_type> idx)
+    {
+        assert(std::size(idx) == self.get_num_of_dimension());
 
-    typename Data::iterator end() { return mData.end(); }
+        std::array<size_type, MaxNumDims> real_idx;
+        for(size_type dim = 0; dim < std::size(idx); ++dim)
+        {
+            auto& slicer  = self.get_slicer(dim);
+            real_idx[dim] = (slicer ? (*slicer)(idx[dim]) : idx[dim]);
+        }
 
-    typename Data::pointer data() { return mData.data(); }
+        return self.mData[self.GetOffsetFromMultiIndex(
+            span<const size_type>(std::data(real_idx), std::size(idx)))];
+    }
 
-    typename Data::const_iterator begin() const { return mData.begin(); }
+    template <typename F>
+    void ForEach_impl(F&& f, std::vector<size_t>& idx, size_t rank)
+    {
+        if(rank == get_num_of_dimension())
+        {
+            f(*this, idx);
+            return;
+        }
+        // else
+        for(size_t i = 0; i < get_length(rank); i++)
+        {
+            idx[rank] = i;
+            ForEach_impl(std::forward<F>(f), idx, rank + 1);
+        }
+    }
 
-    typename Data::const_iterator end() const { return mData.end(); }
+    template <typename F>
+    void ForEach_impl(const F&& f, std::vector<size_t>& idx, size_t rank) const
+    {
+        if(rank == get_num_of_dimension())
+        {
+            f(*this, idx);
+            return;
+        }
+        // else
+        for(size_t i = 0; i < get_length(rank); i++)
+        {
+            idx[rank] = i;
+            ForEach_impl(std::forward<const F>(f), idx, rank + 1);
+        }
+    }
+
+    Data mData;
+    std::array<std::optional<Slicer>, MaxNumDims> mSlicers;
+};
 
-    typename Data::const_pointer data() const { return mData.data(); }
+template <typename T>
+using tensor_view_value_t =
+    remove_cvref_t<decltype(std::declval<remove_cvref_t<T>&>()(0, 0, 0, 0))>;
 
-    typename Data::size_type size() const { return mData.size(); }
+template <typename T>
+struct HostTensor : HostTensorView<T>
```

Comment:
View is higher level data struct.
it is better to make HostTensorView inherit from HostTensor
struct HostTensorView: protected HostTensor

Reply:
We cannot do that because `HostTensor<>` is an owning structure but `HostTensorView<>` is not. 
Indeed they share common interfaces and more importantly: a `HostTensor<>` **is-a** `HostTensorView<>`. In order to model the **is-a** relationship and allow from `HostTensor<>`->`HostTensorView<>` conversions. I decided to make this hierarchy. Otherwise, there will be tons of duplcation codes.